V2 steps

Epigene opened the pull request:
Hi!

I am working on a new version of this gem with these design goals in mind:
1. Allow the gem to take in arbitrary ruby code and produce scaffold for it,
2. Have an easy-to-use CLI like RSpec itself,
3. Add some sugar to scaffolds for Rails Controllers and Models (my primary use-case)

So far I have had good progress with step 1 by updating the Ryan gem.
Step 2 is also progressing well (I can trigger attempts at file processing), however, I am encountering failures that have me wondering how exactly this gem achieves what it does.
Step 3 will come in the following weeks.

The most pressing issue I need clarification with is this: can RSpec::Scaffold work with arbitrary code passed to it? Especially code with `require`, `include` and inheritance lines that refer to code not provided? Or is it feasible only while hooked into actual, working application/gem?

I am looking at `spec/spec_helper:22` and wondering how it would be possible to give specs stand-alone files to process without requiring them.

Epigene committed:
Prepared core test files

Epigene committed:
Prepared core test files

Epigene committed:
Refactored generation expectations out into helper file

Epigene committed:
Refactored generation expectations out into helper file

Epigene committed:
Stashing for ryan v1.1.0 availability

Epigene committed:
Stashing for ryan v1.1.0 availability

Epigene committed:
Started on simple CLI

Epigene committed:
Started on simple CLI

Epigene committed:
Stashing, pending arbitrary code parsing

Epigene committed:
Stashing, pending arbitrary code parsing

Epigene committed:
formalised v2 design goals in changelog

Epigene committed:
formalised v2 design goals in changelog

Epigene committed:
Cleaned exe

Epigene committed:
Cleaned exe

ridiculous committed:
Update Runner and Generator to accept both raw code and files or directories

ridiculous committed:
Update travis rubies

ridiculous committed:
Merge pull request #2 from ridiculous/accept-raw-ruby

Update Runner and Generator to accept both raw code and files or dire…

ridiculous commented:
Hey @Epigene, thanks for the interest 😄 It was kind of a pain to support the raw string, but it's working now, updated in https://github.com/ridiculous/rspec-scaffold/pull/2

As for expanding the functionality, I think it's definitely possible what you're asking. The gem doesn't currently do anything with included/extended modules, but it could be updated, well maybe we'd want to add that to `Ryan`, since we're pretty much just delegating to it. Unfortunately these gems were such a hassle to write 😓 

As for the structure of the Rails specific stuff, I think we should keep that in this repo, as part of the gem, but only require it via a railtie of some sort. There will be a lot of rails specific stuff and we need to clearly draw the boundary.

Please rebase and let me know how it goes!

ridiculous commented:
Maybe add the extra scaffolding for included/extended modules as an optional flag? Now that I think about it, most of those modules functionality is usually tested. I'll try to look into adding `#extended_modules` and `#included_modules` to Ryan, which could be used by the `Runner` to load those files and generate specs for them as well, which can then be zipped into the main spec file (hopefully)

ridiculous commented:
Also, in PR #2 it fixed the issue with loading the test subject

Epigene committed:
Cmon!

Epigene committed:
.testify_text specced

Epigene committed:
Refactored pesky helper file

Epigene commented:
Dat rebase.. :sweat:

I got back on track with the specs, turned out there were lines in spec_helper that were requiring the code in fixtures, breaking things.

Maybe I will have time to move things along over the holiday.

Epigene committed:
TDDed .testify_file

Epigene committed:
Removed pries

Epigene committed:
Cleaned specs

Epigene commented on spec/lib/rspec/scaffold_spec.rb:
```diff
@@ -18,24 +18,73 @@
+
+    context "when passed :to_file mode with no output path" do
+      it "should puts about where test scaffold was saved to, output to a file (in the same directory as original) and return the path" do
+        FakeFS.with_fresh do
```
FakeFS helps with testing changes in filesystem.
It is not without its problems, but seems to work fine here.

Epigene committed:
Prepared CLI core, specs imminent

Epigene commented:
I'm having issues colorizing error messages with Highline. Luckily, I have a five line gem that could work.

It chokes on this as argument

```rb
"  <%= color('- parse error in '/home/augusts/Documents/rspec-scaffold/spec/dummy/app/models/bug.rb': (string):3 :: parse error on value [\"end\", 3] (kEND)', :red) %>\n"
```

Epigene committed:
CLI specs pass

Epigene committed:
Cleaned specs

Epigene commented on lib/rspec/scaffold/cli.rb:
```diff
@@ -16,40 +16,52 @@ def initialize(boot_path)
 
       # This method will be run as a script and uses on ARGV hash contents
       def start
```
CLI is done!

Epigene commented on rspec-scaffold.gemspec:
```diff
@@ -19,9 +19,10 @@ Gem::Specification.new do |spec|
   spec.license       = 'MIT'
 
-  spec.required_ruby_version = '>= 1.9.3'
+  spec.required_ruby_version = '>= 2.0.0'
```
Sadly, my colorizing 'luks' gem has a ruby v2 dep. For now I got around it by bumping rspec-scaffold's req to v2 as well.
How bad do we want to support versions before v2 in upcoming rspec-scaffold version? Seeing how latest version is at 2.4 and officially supported at 2.2, I am very inclined to drop the support.

Epigene commented:
Did a big chunk of work on the CLI. Naively, it should work. Most usefully, calling `rspec-scaffold app/` in a rails project root will ensure a barebones spec file for __all__ ruby files in `/app` dir. :star:

Next up is fleshing out the Generator class with useful tidbits.

Epigene committed:
Had major trouble parsing scope definitions

Epigene committed:
All done

Epigene commented:
Well, here it is, the CLI works and I added some minor touches to model scaffolding, specifically, scaffolding for defined scopes.

I will hold off on controller spec scaffold upgrade because it seems they are being deprecated in favor for request specs, but I am not yet familiar with those.

v2 introduces unfortunate break in backwards compatibility for ruby 1.9.3 with `luks` gem for colored output. I think it's OK to drop the support, seeing how ruby v2.2 if the oldest officially supported at the moment.

ridiculous commented on lib/rspec/scaffold/recursive_directory_expander.rb:
```diff
@@ -0,0 +1,28 @@
+        raise(ArgumentError.new(%Q|"#{file}" is not a directory|)) if !@input_dir.directory?
+
+        # 2. do the recursive expansion of ruby files
+        print ">> Recursively scanning #{@input_dir} for ruby files... "
```
I'm not seeing the recursion here, where is that happening?

Epigene replied:
The name is just fancy.
`Dir.glob("#{@input_dir}**/*.rb")` takes care of looking up all ruby files in subdirecories.

ridiculous replied:
Ok... so do we really need this class then? I do appreciate the error handling. I'd be fine with just renaming to `DirExpander`

Epigene replied:
Will do.

ridiculous commented on README.md:
```diff
@@ -1,22 +1,74 @@
+The idea is to point to existing ruby code files and `rspec-scaffold` will ensure corresponding spec files in `/spec` directory.  
+
+### The CLI
+Only operates on paths.
```
What does this mean? Kinda confusing, cause line 36 we pass it a _file_, not a path

Epigene replied:
It is supposed to mean that CLI only takes files/paths as arguments, not raw ruby code.

ridiculous replied:
ok, cool, let's mention that here

ridiculous commented on README.md:
````diff
@@ -1,22 +1,74 @@
+__3. Provide ruby code file(s) -> get scaffold file(s)__
+
+```rb
+RSpec::Scaffold.testify_file(filepath, :to_file, "/optional/custom/output/file.rb")  
````
I like the API for the `system` method, which it takes an optional `:out` keyword argument that redirects output from STDOUT to a file. How do you feel about copying that here? If the `:out` keyword is supplied, then we assume it's `:to_file`, otherwise return text, and can remove the `mode` argument

Epigene replied:
Excellent point! Will refactor.

Epigene replied:
On the second thought, a pattern like that can not support the default use case - outputting to default file inferred from input file location.

Epigene replied:
On the third thought, optional `:out` keyword keyword argument, when passed can work as if `:to_file nad 'path/to/file'` were both passed, without affecting current logic.

ridiculous replied:
ok, sounds good!

ridiculous commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
-  - 2.2.2
-  - 2.2.3
+  - 2.2.6
+  - 2.4.0
```
Can we add 2.3 to this list and drop 1.9.3?

Epigene replied:
Yes, I think everything after 2.1 will definitely work, and 2.0 will possibly work.

ridiculous reviewed (commented):
This is really awesome!! Great work, I really like it :clap: :clap:

Epigene committed:
Finalizing refactors

Epigene commented:
What I did:
1) Renamed `RecursiveDirectoryExpander` to `DirExpander`,  
2) Clarified Readme as to what arguments CLI accepts,
3) Refactored `.testify_file` to not need mode if output file keyword argument is passed,  
4) Updated Travis to cover latest, stable releases of ruby versions 2.1.* through 2.4.*.

ridiculous commented:
Excellent, thank you! I'll try it out tonight and merge if it's all good 🚀

ridiculous commented:
We need to move "pry" to a development dependency

ridiculous commented:
I'm seeing two slashes in file outputs from `rspec-scaffold app/models`. Looks like:
```bash
- /spec/models//user_spec.rb - already exists
+ /spec/models//user_session_spec.rb
```
You know what's causing that?

ridiculous commented:
I get a permission error running `rspec-scaffold lib/app.rb` (or any single file in the lib directory): https://gist.github.com/ridiculous/43c484b51d91882289da50ebec9e006a

Epigene committed:
Fixed bugs found by Ryan

Epigene commented:
Thanks for going over it, Ryan!

1) Moved pry back to development deps
2) Fixed the two slash bug, was an issue with Dir.glob argument
3) Permission error due to bug in spec location builder, added spec and fix.

Verdict - Im a newb when it comes to working with `Filepath` class :sweat_smile:

ridiculous commented:
Looks good! Nice work :+1:

Epigene commented:
Yayifications!
